### PR TITLE
Customizable `LiveView` state views

### LINKED ISSUE #1232: Customizable `LiveView` state indicator Views
Currently, customizing the loading, error, etc. states of a `LiveView` requires a `RootRegistry`. This API could be simplified by providing extra `ViewBuilder` closures on the `LiveView` initializer:

```swift
LiveView(.localhost) {
  ProgressView("Loading")
} error: { error in
  #if DEBUG
  Text(error.localizedDescription)
  #else
  Text("An error occurred")
  #endif
} reconnecting: { content in
  content
    .overlay {
      Text("Reconnecting")
    }
} disconnected: {
  Text("Failed to connect")
}
```

These arguments would be given priority over the `RootRegistry`. If they are not provided, the Views in the `RootRegistry` will be used.

### PR DESCRIPTION
**Pull request by carson-katri**
Closes #1232 

State views can be provided as arguments to the `LiveView`. The `loadingView` method was from the `CustomRegistry` protocol, as it makes more sense to provide these options on the `LiveView` itself. Here's an example of using these arguments:

```swift
LiveView(.localhost) {
    ProgressView("Loading")
} disconnected: {
    Text("Not connected to the server")
} reconnecting: { content in
    content
        .safeAreaInset(edge: .bottom) {
            HStack {
                ProgressView()
                Text("Reconnecting")
            }
            .padding()
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Color.orange)
        }
} error: { error in
    Text(error.localizedDescription)
}
```

Note that `reconnecting` provides the `content` of the `LiveView`, allowing you to create an overlay instead of fully replacing the screen when the connection temporarily drops.

All of these arguments are optional. If any is omitted, a default View will be used. They can also be used with the `#LiveView` macro when passing `addons`.

A new `registry` argument was also added. Previously, registries were passed as generic arguments:
```swift
LiveView<MyRegistryType>(.localhost)
```
With the addition of these state views, there are more generic arguments that must be provided.
```swift
LiveView<MyRegistryType, _, _, _, _>(.localhost)
```
Instead of passing a generic argument, it is now recommended to pass the `registry` argument:
```swift
LiveView(registry: MyRegistryType.self, .localhost)
```
However, it is generally easier to use the `#LiveView` macro to pass custom registries, so this argument is almost never required unless some further customization is required:
```swift
#LiveView(.localhost, addons: [MapKitRegistry<_>.self, ChartsRegistry<_>.self])
```

### CODE REVIEW

**Comment by bcardarella** [comment]
I'm going to go through the js code and just verify this represents all of the states we could go into

**Comment by carson-katri** [comment]
`reconnect` states can now be rendered:

```swift
LiveView(.localhost) {
    ProgressView("Loading")
} reconnecting: { content in
    content
        .safeAreaInset(edge: .bottom) {
            HStack {
                ProgressView()
                Text("Reconnecting")
            }
            .padding()
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Color.orange)
        }
}
```

In this video, I killed the server while the app was running, then started it again a few seconds later:

https://github.com/liveview-native/liveview-client-swiftui/assets/13581484/f9f4fca8-8656-4760-8844-e4048aaa6fa5

**Comment by carson-katri** [comment]
A default `reconnecting` is provided:

https://github.com/liveview-native/liveview-client-swiftui/assets/13581484/8fcb9fa1-e699-4565-845d-758977faf67a


https://github.com/liveview-native/liveview-client-swiftui/assets/13581484/484a1d65-31d4-49fd-a3c1-2a7323137510

**Comment by bcardarella** [comment]
This is excellent!

**Review by bcardarella — APPROVED** [review]

File: Sources/LiveViewNative/NavStackEntryView.swift [hunk 1]
```diff
@@ -43,23 +43,19 @@ struct NavStackEntryView<R: RootRegistry>: View {
...
+                        case .connected, .reconnecting:
+                            fatalError()
+                        case .notConnected:
+                            SwiftUI.Text("Not Connected")
+                        case .connecting:
+                            SwiftUI.ProgressView("Connecting")
+                        case .connectionFailed(let error):
+                            SwiftUI.VStack {
+                                SwiftUI.Text("Connection Failed")
+                                    .font(.subheadline)
```
Comment (bcardarella) [inline Sources/LiveViewNative/NavStackEntryView.swift:56]: I think we should remove this from being default, and that all default states should be `nil`. Instead we should move these into the template project generator so people can have direct access to modify.

**Review by bcardarella — CHANGES_REQUESTED** [review]

File: Sources/LiveViewNative/NavStackEntryView.swift (hunk 1 above)
Comment (carson-katri) [reply Sources/LiveViewNative/NavStackEntryView.swift:56]: So should the state view arguments be required? Or do states like `connecting`, `disconnected`, etc. just render a blank screen if no argument is passed?

Comment (bcardarella) [reply Sources/LiveViewNative/NavStackEntryView.swift:56]: No, not required. If they aren't provided it's just a `no-op`

Comment (carson-katri) [reply Sources/LiveViewNative/NavStackEntryView.swift:56]: How about the LiveView error webpage View. Should we display that by default?

Comment (bcardarella) [reply Sources/LiveViewNative/NavStackEntryView.swift:56]: oh that's a good point, how about the view is shipped with the client but we still have the code to render that view in the error closure generated into the template

**Comment by carson-katri** [comment]
@bcardarella Here are the final template states:

<table>
<tr><th>Connecting</th><th>Reconnecting</th><th>Error (Release Mode)</th><th>Server Error (Debug Mode)</th><th>Connection Error (Debug Mode)</th></tr>
<tr>
<td>



</td>
<td>



</td>
<td>

Simulator Screenshot - iPhone 15 Pro - 2024-01-31 at 13 57 06

</td>
<td>

Simulator Screenshot - iPhone 15 Pro - 2024-02-01 at 14 42 26

</td>
<td>

Simulator Screenshot - iPhone 15 Pro - 2024-02-01 at 14 40 29

</td>
</tr>
</table>

File: priv/xcodegen/Sources/TemplateApp/ContentView.swift [hunk 2]
```diff
@@ -8,9 +8,85 @@ import LiveViewNative
...
-            development: .localhost(path: "%LVN_PREFERRED_ROUTE%"),
-            production: .custom(URL(string: "https://%LVN_PREFERRED_PROD_URL%%LVN_PREFERRED_ROUTE%")!)
-        ))
+        #LiveView(
+            .automatic(
+                development: .localhost(path: "%LVN_PREFERRED_ROUTE%"),
+                production: URL(string: "https://%LVN_PREFERRED_PROD_URL%%LVN_PREFERRED_ROUTE%")!
+            ),
+            addons: []
+        ) {
```
Comment (carson-katri) [inline priv/xcodegen/Sources/TemplateApp/ContentView.swift:17]: The template uses the `#LiveView` macro with an empty `addons` parameter by default now.

**Review by bcardarella — APPROVED** [review]

**Pull request merged**